radicle-crypto/src/ssh/agent.rs

```diff
@@ -1,21 +1,92 @@
-use radicle_ssh::agent::client::AgentClient;
-use radicle_ssh::{self as ssh, agent::client::ClientStream};
+use std::ops::{Deref, DerefMut};
+use std::sync::Mutex;
+
+pub use radicle_ssh::agent::client::AgentClient;
+pub use radicle_ssh::agent::client::Error;
+pub use radicle_ssh::{self as ssh, agent::client::ClientStream};
 
-use crate as crypto;
 use crate::ssh::SecretKey;
+use crate::{self as crypto};
+use crate::{PublicKey, Signature, Signer, SignerError};
 
 #[cfg(not(unix))]
-use std::net::TcpStream as Stream;
+pub use std::net::TcpStream as Stream;
 #[cfg(unix)]
-use std::os::unix::net::UnixStream as Stream;
+pub use std::os::unix::net::UnixStream as Stream;
+
+pub struct Agent {
+    client: AgentClient<Stream>,
+}
+
+impl Agent {
+    /// Connect to a running SSH agent.
+    pub fn connect() -> Result<Self, ssh::agent::client::Error> {
+        Stream::connect_env().map(|client| Self { client })
+    }
+
+    /// Register a key with the agent.
+    pub fn register(&mut self, key: &crypto::SecretKey) -> Result<(), ssh::Error> {
+        self.client.add_identity(&SecretKey::from(*key), &[])
+    }
+}
+
+impl Deref for Agent {
+    type Target = AgentClient<Stream>;
 
-pub fn connect() -> Result<AgentClient<Stream>, ssh::agent::client::Error> {
-    Stream::connect_env()
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
 }
 
-pub fn register(key: &crypto::SecretKey) -> Result<(), ssh::agent::client::Error> {
-    let mut agent = self::connect()?;
-    agent.add_identity(&SecretKey::from(*key), &[])?;
+impl DerefMut for Agent {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.client
+    }
+}
+
+/// A [`Signer`] that uses `ssh-agent`.
+pub struct AgentSigner {
+    agent: Mutex<Agent>,
+    public: PublicKey,
+}
+
+impl AgentSigner {
+    pub fn new(agent: Agent, public: PublicKey) -> Self {
+        let agent = Mutex::new(agent);
+
+        Self { agent, public }
+    }
+
+    pub fn is_ready(&self) -> Result<bool, Error> {
+        let ids = self.agent.lock().unwrap().request_identities()?;
+
+        Ok(ids.contains(&self.public))
+    }
+
+    /// Box this signer into a [`Signer`].
+    pub fn boxed(self) -> Box<dyn Signer> {
+        Box::new(self)
+    }
+}
+
+impl Signer for AgentSigner {
+    fn public_key(&self) -> &PublicKey {
+        &self.public
+    }
+
+    fn sign(&self, msg: &[u8]) -> Signature {
+        self.try_sign(msg).unwrap()
```

Comment:
expect() may be better?

Reply:
Yeah, I kind of go back and forth on `expect` -- I feel that here it's pretty obvious that we expect try_sign not to fail and so the expect string feels redundant..

Reply:
Would the panic still be clear to operators if it did explode somewhere?